lib/srv/usermgmt.go

```diff
@@ -257,28 +256,44 @@ func (u *HostUserManagement) updateUser(name string, ui services.HostUsersInfo)
 
 	_, hasDropGroup := currentGroups[types.TeleportDropGroup]
 	_, hasKeepGroup := currentGroups[types.TeleportKeepGroup]
-	if !hasDropGroup && !hasKeepGroup && !migrateKeepUser {
-		return trace.Errorf("%q %w", name, unmanagedUserErr)
+	_, hasStaticGroup := currentGroups[types.TeleportStaticGroup]
+	if !(hasDropGroup || hasKeepGroup || hasStaticGroup || migrateKeepUser) {
+		return nil, trace.Errorf("%q %w", name, unmanagedUserErr)
+	}
+
+	// Do not convert/update groups from static to non-static user, and vice versa.
+	if hasStaticGroup != ui.IsStaticUser {
+		return nil, nil
 	}
 
+	var closer io.Closer
 	switch ui.Mode {
 	case types.CreateHostUserMode_HOST_USER_MODE_INSECURE_DROP:
 		ui.Groups = append(ui.Groups, types.TeleportDropGroup)
+		closer = &userCloser{
+			username: name,
+			users:    u,
+			backend:  u.backend,
+		}
 	case types.CreateHostUserMode_HOST_USER_MODE_KEEP:
-		if !hasKeepGroup {
-			home, err := u.backend.GetDefaultHomeDirectory(name)
-			if err != nil {
-				return trace.Wrap(err)
-			}
+		if hasStaticGroup {
+			ui.Groups = append(ui.Groups, types.TeleportStaticGroup)
```

Comment:
Since this is the same path as non-static host user creation, this means a static user's group assignments can be modified if you have a role that allows logging in as that user. Is that something that we should allow? It seems like you might want a static user to always have a particular set of groups assigned, regardless of whether or not a user starts a session against it.

Reply:
Groups will only be updated if both the existing user and mode are static. I think the shared keep mode was obscuring that, it should be clearer now with the internal enum.